### ISSUE TITLE
bang syntax regression

### ISSUE THREAD

**Issue by gforsyth** [xonsh/xonsh#1723]
Somewhere along the way, `!!` and `!n` broke.  

``` console
gil@theo ~/tmp $ ls
Dripppy  Drippy  doctr  test  testgit
gil@theo ~/tmp $ !!<xonsh-code>:1:0 - !!      
<xonsh-code>:1:0 + ![!!]
gil@theo ~/tmp $ !!
................ 
<xonsh-code>:1:0 - !!
<xonsh-code>:1:0 + ![!!]
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/ptk/shell.py", line 99, in push
    locs=None)
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/execer.py", line 103, in compile
    transform=transform)
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/execer.py", line 73, in parse
    tree, input = self._parse_ctx_free(input, mode=mode, filename=filename)
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/execer.py", line 177, in _parse_ctx_free
    raise original_error from None
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/execer.py", line 164, in _parse_ctx_free
    debug_level=(self.debug_level > 1))
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/parsers/base.py", line 328, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/home/gil/anaconda/lib/python3.5/site-packages/ply/yacc.py", line 265, in parse
    return self.parseopt_notrack(input,lexer,debug,tracking,tokenfunc)
  File "/home/gil/anaconda/lib/python3.5/site-packages/ply/yacc.py", line 1047, in parseopt_notrack
    tok = self.errorfunc(errtoken)
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/parsers/base.py", line 2691, in p_error
    column=p.lexpos))
  File "/home/gil/anaconda/lib/python3.5/site-packages/xonsh/parsers/base.py", line 459, in _parse_error
    raise err
  File "<string>", line None
SyntaxError: <xonsh-code>:1:0: ('code: !',)
!!
^
```

**Comment by scopatz** [comment]
Hi @gforsyth, sorry about this. I anticipated that this might happen.  

I think that the fix is to have a bashisms xontrib that implements a precommand event handle which does does the textual history replacement.  This has the advantage of being more like what bash itself does, where you can use !! anywhere in the line, and not just as an alias.  This would also let othere bashisms naturally come in too.

Sorry that this broke :(

**Comment by gforsyth** [comment]
No worries!  And yeah, best to remove them rather than leaving broken things behind.  

I can look at putting together a bashisms xontrib if I really end up missing `!!`

**Comment by scopatz** [comment]
If you don't mind, I think we should leave this open so that other people know will know it is potentially in the works... or they might beat us to it :)

**Comment by gforsyth** [comment]
This was addressed (and then some) by #1968
